compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -938,6 +940,36 @@ impl<'a> Parser<'a> {
             && self.look_ahead(1, |tok| tok.kind == TokenKind::OpenDelim(Delimiter::Parenthesis))
             && let Some(path) = self.recover_path_from_fn()
         {
+            path
+        } else if !self.token.is_path_start() && self.token.can_begin_type() {
+            // Instead of finding a path (a trait), we found a type.
+            let mut err = self.struct_span_err(self.token.span, "expected a trait, found type");
```

Comment:
```suggestion
            let mut err = self.struct_span_err(self.token.span, "expected a trait, found type start");
```
Specify "type start" because we *haven't* parsed a type yet, and the span points at `&` only, so it would be a "lie" (and potentially confusing) to talk about a type.

Reply:
I've done this slightly differently. Instead we will only emit the nice error is there is a proper type, "found type start" seems a bit confusing to me.